cli/command/image/build.go

```diff
@@ -227,13 +239,56 @@ func runBuild(dockerCli *command.DockerCli, options buildOptions) error {
 		buildCtx, err = archive.TarWithOptions(contextDir, &archive.TarOptions{
 			Compression:     compression,
 			ExcludePatterns: excludes,
-			IncludeFiles:    includes,
 		})
 		if err != nil {
 			return err
 		}
 	}
 
+	// replace Dockerfile if added dynamically
+	if dockerfileCtx != nil {
+		file, err := ioutil.ReadAll(dockerfileCtx)
+		dockerfileCtx.Close()
+		if err != nil {
+			return err
+		}
+		now := time.Now()
+		hdrTmpl := &tar.Header{
+			Mode:       0600,
+			Uid:        0,
+			Gid:        0,
+			ModTime:    now,
+			Typeflag:   tar.TypeReg,
+			AccessTime: now,
+			ChangeTime: now,
+		}
+		randomName := ".dockerfile." + stringid.GenerateRandomID()[:20]
+
+		buildCtx = archive.ReplaceFileTarWrapper(buildCtx, map[string]archive.TarModifierFunc{
+			randomName: func(_ string, h *tar.Header, content io.Reader) (*tar.Header, []byte, error) {
+				return hdrTmpl, file, nil
+			},
+			".dockerignore": func(_ string, h *tar.Header, content io.Reader) (*tar.Header, []byte, error) {
+				if h == nil {
+					h = hdrTmpl
+				}
+				extraIgnore := randomName + "\n"
+				b := &bytes.Buffer{}
+				if content != nil {
+					_, err := b.ReadFrom(content)
+					if err != nil {
+						return nil, nil, err
+					}
+				} else {
+					extraIgnore += ".dockerignore\n"
+				}
+				b.Write([]byte("\n" + extraIgnore))
```

Comment:
I've changed the logic here to write to the buffer directly instead of using the `extraIgnore` string. I think the idea here was that you don't know if their content ended with a newline, so you have to add another here to be sure.